File: go/142.Linked_List_Cycle_II/linkedListCycleII.go
func detectCycle(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
        return nil
    }
    
    slow, fast := head, head
    for fast != nil {
        slow = slow.Next
        if fast.Next == nil { break }
        fast = fast.Next.Next
        if slow == fast { break }
    }
    // no cycle
    if fast == nil || fast.Next == nil {
        return nil
    }
    
    slow, fast = head, slow
    for slow != fast {
        slow = slow.Next
        fast = fast.Next
    }
    
    return slow
}

File: python/142.Linked_List_Cycle_II/linkedListCycleII.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def detectCycle(self, head):
    	while head is not None:
    		if isinstance(head.val, str):
    			return head
    		head.val = str(head.val)
    		head = head.next
    	return None
    def detectCycle2(self,head):
    	tag = {}
    	while head is not None:
    		tag[head] = 1
    		head = head.next
    		if head in tag:
    			return head
    	return None
